chainermn/optimizers.py

```diff
@@ -46,15 +51,117 @@ def __setattr__(self, attr_name, value):
         setattr(self.actual_optimizer, attr_name, value)
 
 
-def create_multi_node_optimizer(actual_optimizer, communicator):
+class _DoubleBufferingOptimizer(object):
+
+    def __init__(self, actual_optimizer, communicator):
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'communicator', communicator)
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'actual_optimizer', actual_optimizer)
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'needs_update', False)
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'device', None)
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'thread_pool', multiprocessing.pool.ThreadPool(1))
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'communicated_target', None)
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'target_params_list', [[], []])
+        super(_DoubleBufferingOptimizer, self).__setattr__(
+            'allreduce_grad_res', None)
```

Comment:
What does `res` mean? I could not understand only from the code.

Reply:
`res` means the result  of allreduce_grad. I will rename it to a more descriptive name.